django/db/models/query.py

```diff
@@ -223,6 +223,9 @@ def __or__(self, other):
         combined.query.combine(other.query, sql.OR)
         return combined
 
+    def __contains__(self, obj):
+        return isinstance(obj, self.model) and self.filter(pk=obj.pk).exists()
```

Comment:
Fair enough - I'll change it to use the cached results, which, you're right, it should probably do.

However - if anyone's currently relying on `__contains__` to evaluate a QuerySet, they probably shouldn't. It's a little-known side effect and causes a lot more harm than good. The Pythonic way of testing membership is `in` - every Python devs knows this and will use it naturally. The fact that using `in` causes an *enormous* performance hit on list-like QuerySets is something most devs won't see coming, and in a lot of cases won't notice until it's far too late.

This could be held up till a major release - 1.9 or 2.0 - where devs could expect changes to core behaviour. But I'm strongly of the opinion that, sooner or later, it should be known that you can use `in` for testing membership like you would any other Python collection, and that it has no side effects on your QuerySet.